jq255.md

```diff
@@ -0,0 +1,1412 @@
+    implemented by a right-shift by 1 bit, then the conditional addition
+    of the constant `(q+1)/2` if the dropped bit was non-zero. Such an 
+    implementation is considered to be a "fast" operation (i.e. at least
+    as fast as an addition in the field).
+
+  - `invert(x)` returns the multiplicative inverse of `x` in the field.
+    Zero has no multiplicative inverse, but it is sometimes convenient
+    for constant-time implementations to return `0` as a (formal) inverse
+    of `0`; this is obtained naturally if inversion is implemented using
+    Fermat's Little Theorem (i.e. raising `x` to the power `q-2`). This
+    feature is not needed for the formulas described in this document.
+
+  - `sgn(x)` returns the "sign" of `x`, as defined previously: the
+    sign is the value of the least significant bit of `int(x)`.
+
+  - `sqrt(x)` returns either a square root of `x` (if `x` is indeed
+    a square in the finite field), or `False` (if `x` is not a square
+    in the finite field). **Important:** when `x` is a square, the
+    returned square root is non-negative.
+
+  - `is_square(x)` returns `True` if `x` is a square in the finite field,
+    or `False` otherwise.
+
+  - `scalar_to_int(s)` and `int_to_scalar(j)` compute the same
+    operations as `int()` and `GFq()`, respectively, but for scalars,
+    which are integers modulo the prime `r`.
+
+  - `secure_random(m)`, for an integer `m >= 1`, returns a sequence of
+    `m` random bytes obtained from a cryptographically secure source,
+    with a uniform probability distribution (e.g. `/dev/urandom` or the
+    `getrandom()` system call on many Unix-like operating syustems).
+    This is used only for private key generation in the specification of
+    high-level protocols.
+
+If an addition, subtraction or multiplication uses as operands a field
+element and an integer (usually small), then the integer is first
+implicitly converted to a field element as if by using `GFq()`. In
+practice, the implementation may provide a dedicated
+multiply-by-small-integer function; on usual software platforms, such an
+operation is about as fast as an addition in the field.
+
+The implementation of `sqrt(x)` depends on the used field. For jq255s,
+the field is such that `q = 3 mod 4`, which allows a simple square root
+implementation by computing the square root candidate `z = x^((q+1)/4)`.
+For jq255e, the field is such that `q = 5 mod 8`, which can be handled
+with Atkin's algorithm:
+
+ 1. Compute: `c = (2*x)^((q-5)/8)`
+ 2. Compute: `d = 2*x*c^2`
+ 3. Compute the square root candidate: `z = x*c*(d - 1)`
+
+The obtained candidate MUST be checked (by comparing `z^2` with `x`).
+Moreover, if `z` is indeed a square root of `x`, then it MUST be
+normalized to a non-negative value (i.e. if `sgn(z) = 1`, then `z` must
+be replaced with `-z`). The use of this fixed convention for square
+roots is needed for proper interoperability of the map-to-group and
+hash-to-group operations.
```

Comment:
Should we bring in RFC 2119 for 'MUST' etc?

Reply:
It depends on how much you want C2SP to mimic or fight the IETF.

Reply:
Heh, neither is the goal, really. There are some things of the IETF culture that are good. RFC 2119, I'd argue, is at least good by being universally known by the folks who implement specs. There's an argument that maybe there shouldn't be that many SHOULD in a robust spec, but in general I am not against using MUST/SHOULD/etc and referencing RFC 2119 if that's how we are using them.

Reply:
I have added a reference to RFC 2119 in the "conventions" section.

Reply:
🙏✔️